src/middleware/routeHandler.ts

```diff
@@ -0,0 +1,250 @@
+  try {
+    await lock(siteName)
+  } catch (err) {
+    next(err)
+    return
+  }
+
+  let originalStagingCommitSha: any
+  let originalStagingLiteCommitSha: any
+
+  if (shouldUseGitFileSystem) {
+    const results = await Promise.all([
+      gitFileSystemService.getLatestCommitOfBranch(siteName, STAGING_BRANCH),
+      gitFileSystemService.getLatestCommitOfBranch(
+        siteName,
+        STAGING_LITE_BRANCH
+      ),
+    ])
+    const [stagingResult, stagingLiteResult] = results
+
+    if (stagingResult.isErr() || stagingLiteResult.isErr()) {
+      await unlock(siteName)
+      if (stagingResult.isErr()) next(stagingResult.error)
+      if (stagingLiteResult.isErr()) next(stagingLiteResult.error)
+      return
+    }
+    originalStagingCommitSha = stagingResult.value.sha
+    originalStagingLiteCommitSha = stagingLiteResult.value.sha
+    if (!originalStagingCommitSha || !originalStagingLiteCommitSha) {
+      await unlock(siteName)
+      if (stagingResult.isErr()) next(stagingResult.error)
+      return
+    }
+    // Unused for git file system, but to maintain existing structure
+    res.locals.githubSessionData = new GithubSessionData({
+      currentCommitSha: "",
+      treeSha: "",
+    })
+  } else {
+    try {
+      const {
+        currentCommitSha: currentStgCommitSha,
+        treeSha: stgTreeSha,
+      } = await getCommitAndTreeSha(siteName, accessToken, STAGING_BRANCH)
+
+      const {
+        currentCommitSha: currentStgLiteCommitSha,
+      } = await getCommitAndTreeSha(siteName, accessToken, STAGING_LITE_BRANCH)
+
+      const githubSessionData = new GithubSessionData({
+        currentCommitSha: currentStgCommitSha,
+        treeSha: stgTreeSha,
+      })
+      res.locals.githubSessionData = githubSessionData
+
+      originalStagingCommitSha = currentStgCommitSha
+      originalStagingLiteCommitSha = currentStgLiteCommitSha
+    } catch (err) {
+      await unlock(siteName)
+      next(err)
+      return
+    }
+  }
+
+  await routeHandler(req, res, next).catch(async (err: any) => {
+    try {
+      if (shouldUseGitFileSystem) {
+        await backOff(() =>
+          convertNeverThrowToPromise(
+            gitFileSystemService.rollback(
+              siteName,
+              originalStagingCommitSha,
+              STAGING_BRANCH
+            )
+          )
+        )
+
+        await backOff(() =>
+          convertNeverThrowToPromise(
+            gitFileSystemService.rollback(
+              siteName,
+              originalStagingLiteCommitSha,
+              STAGING_LITE_BRANCH
+            )
+          )
+        )
+
+        await backOff(() => {
+          let pushRes = gitFileSystemService.push(
+            siteName,
+            STAGING_BRANCH,
+            true
+          )
+          if (originalStagingLiteCommitSha) {
+            pushRes = pushRes.andThen(() =>
+              gitFileSystemService.push(siteName, STAGING_LITE_BRANCH, true)
+            )
+          }
```

Comment:
given that this is taken mostly as-is, it's ok to keep with the existing structure. just thinking out loud here, no real action expected on your end - why wouldn't we chain the results together then do a single `backOff` at the end?

Reply:
there might be semantic differences and i'm not entirely sure the existing impl is correct here but given that the structure is kept, i'm not too keen on rewriting this now.